repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala

```diff
@@ -44,20 +42,34 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
 
   private[repl] var session: Session = _
 
+  private val kind = Kind(livyConf.get(RSCConf.Entry.SESSION_KIND))
+
+  private var pysparkJobProcessor:PySparkJobProcessor = null
+
   override protected def initializeContext(): JavaSparkContext = {
-    val interpreter = Kind(livyConf.get(RSCConf.Entry.SESSION_KIND)) match {
+    val interpreter = kind match {
       case PySpark() => PythonInterpreter(conf, PySpark())
       case PySpark3() => PythonInterpreter(conf, PySpark3())
       case Spark() => new SparkInterpreter(conf)
       case SparkR() => SparkRInterpreter(conf)
     }
-
     session = new Session(interpreter)
+    initiatePy4jCallbackGateway(PythonInterpreter.getGatewayServer())
```

Comment:
Why is this here and not in `PythonInterpreter`? Then you wouldn't need the `gatewayServer` variables in the PythonInterpreter object; everything would be handled internally in the PythonInterpreter instance.